### PR TITLE
Feat - Caching layer to database

### PR DESCRIPTION
**Pull request by kodumbeats**
This PR allows for the use of a cache when creating a database.

### CODE REVIEW

**Comment by kodumbeats** [comment]
I can't seem to decode Documents from cache correctly - got any advice?
```bash
Time: 00:00.301, Memory: 6.00 MB

There were 3 errors:

1) Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteAttribute
Exception: Must define $id attribute

/usr/src/code/src/Database/Database.php:548
/usr/src/code/src/Database/Database.php:291
/usr/src/code/tests/Database/Base.php:51

2) Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteIndex
Exception: Must define $id attribute

/usr/src/code/src/Database/Database.php:548
/usr/src/code/src/Database/Database.php:291
/usr/src/code/tests/Database/Base.php:99

3) Utopia\Tests\Adapter\MariaDBTest::testCreateDocument
Exception: Must define $id attribute

/usr/src/code/src/Database/Database.php:548
/usr/src/code/src/Database/Database.php:291
/usr/src/code/tests/Database/Base.php:127

--

There was 1 failure:

1) Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteCollection
Failed asserting that true matches expected false.

/usr/src/code/tests/Database/Base.php:42
```

**Comment by kodumbeats** [comment]
```
PHPUnit 9.5.4 by Sebastian Bergmann and contributors.

..EEESSS..SSSSSS                                                  16 / 16 (100%)

Time: 00:01.354, Memory: 6.00 MB

There were 3 errors:

1) Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteAttribute
json_decode() expects parameter 1 to be string, array given

/usr/src/code/src/Database/Database.php:825
/usr/src/code/src/Database/Database.php:482
/usr/src/code/src/Database/Database.php:234
/usr/src/code/src/Database/Database.php:278
/usr/src/code/tests/Database/Base.php:51

2) Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteIndex
json_decode() expects parameter 1 to be string, array given

/usr/src/code/src/Database/Database.php:825
/usr/src/code/src/Database/Database.php:482
/usr/src/code/src/Database/Database.php:234
/usr/src/code/src/Database/Database.php:278
/usr/src/code/tests/Database/Base.php:99

3) Utopia\Tests\Adapter\MariaDBTest::testCreateDocument
json_decode() expects parameter 1 to be string, array given

/usr/src/code/src/Database/Database.php:825
/usr/src/code/src/Database/Database.php:482
/usr/src/code/src/Database/Database.php:234
/usr/src/code/src/Database/Database.php:278
/usr/src/code/tests/Database/Base.php:127

ERRORS!
Tests: 16, Assertions: 9, Errors: 3, Skipped: 9.

```

**Comment by kodumbeats** [comment]
In 66c9f70, I discovered the following behavior, which has been causing our issues:
```php
$document = $this->adapter->getDocument($collection->getId(), $id);
$document !== new Document($document->getArrayCopy())
```
https://gist.github.com/kodumbeats/d910b12983540bdb125219b7791f27fb

**Comment by kodumbeats** [comment]
As of 53d521c:
```
$ docker-compose exec tests vendor/bin/phpunit --debug tests/Database/Adapter/MariaDBTest.php
PHPUnit 9.5.4 by Sebastian Bergmann and contributors.

Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDelete' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDelete' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteCollection' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteCollection' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteAttribute' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteAttribute' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteIndex' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDeleteIndex' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDocument' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testCreateDocument' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testGetDocument' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testGetDocument' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testUpdateDocument' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testUpdateDocument' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testDeleteDocument' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testDeleteDocument' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testFindFirst' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testFindFirst' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testFindLast' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testFindLast' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testReadPermissionsSuccess' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testReadPermissionsSuccess' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testReadPermissionsFailure' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testReadPermissionsFailure' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testWritePermissionsSuccess' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testWritePermissionsSuccess' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testWritePermissionsFailure' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testWritePermissionsFailure' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testWritePermissionsUpdateFailure' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testWritePermissionsUpdateFailure' ended
Test 'Utopia\Tests\Adapter\MariaDBTest::testExceptionDuplicate' started
Test 'Utopia\Tests\Adapter\MariaDBTest::testExceptionDuplicate' ended


Time: 00:00.737, Memory: 6.00 MB

OK (16 tests, 94 assertions)
```

File: src/Database/Database.php [hunk 1]
```diff
@@ -544,9 +576,11 @@ public function updateDocument(string $collection, string $id, Document $documen
         }
 
         $document = $this->adapter->updateDocument($collection->getId(), $document);
-        
         $document = $this->decode($collection, $document);
 
+        $this->cache->purge($id);
+        $this->cache->save($id, $document->getArrayCopy());
```
Comment (eldadfux) [inline src/Database/Database.php:582]: We need to save the "new" document ID in case it was changed

Comment (kodumbeats) [reply src/Database/Database.php]: Resolved in 2350e5c

**Pull request merged**